### ISSUE TITLE
[Apple Silicon M1 MPS device] bad performance metrics for BERT model training

### ISSUE THREAD

**Issue by pacman100** [pytorch/pytorch#82707]
### 🐛 Describe the bug

When using `mps` device, BERT finetuning on MRPC task leads to bad performance metrics in comparison to CPU training. Also, speedup is only ~30% when compared to CPU training.

### Steps to reproduce
1. Code is given below. The is run on a Mac Pro with M1 chip having 8 CPU performance cores (+2 efficiency cores), 14 GPU cores and 16GB of unified memory.

```python

import argparse

import torch
from torch.optim import AdamW
from torch.utils.data import DataLoader

import evaluate
from accelerate import Accelerator, DistributedType
from datasets import load_dataset
from transformers import AutoModelForSequenceClassification, AutoTokenizer, get_linear_schedule_with_warmup, set_seed



MAX_GPU_BATCH_SIZE = 16
EVAL_BATCH_SIZE = 32

def get_dataloaders(batch_size: int = 16):
    """
    Creates a set of `DataLoader`s for the `glue` dataset,
    using "bert-base-cased" as the tokenizer.

    Args:
        batch_size (`int`, *optional*):
            The batch size for the train and validation DataLoaders.
    """
    tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")
    datasets = load_dataset("glue", "mrpc")

    def tokenize_function(examples):
        # max_length=None => use the model max length (it's actually the default)
        outputs = tokenizer(examples["sentence1"], examples["sentence2"], truncation=True, max_length=None)
        return outputs

    tokenized_datasets = datasets.map(
        tokenize_function,
        batched=True,
        remove_columns=["idx", "sentence1", "sentence2"],
    )

    # We also rename the 'label' column to 'labels' which is the expected name for labels by the models of the
    # transformers library
    tokenized_datasets = tokenized_datasets.rename_column("label", "labels")

    def collate_fn(examples):
        
        return tokenizer.pad(examples, padding="longest", return_tensors="pt")

    # Instantiate dataloaders.
    train_dataloader = DataLoader(
        tokenized_datasets["train"], shuffle=True, collate_fn=collate_fn, batch_size=batch_size
    )
    eval_dataloader = DataLoader(
        tokenized_datasets["validation"], shuffle=False, collate_fn=collate_fn, batch_size=EVAL_BATCH_SIZE
    )

    return train_dataloader, eval_dataloader


def training_function(config, device):
    # Sample hyper-parameters for learning rate, batch size, seed and a few other HPs
    lr = config["lr"]
    num_epochs = int(config["num_epochs"])
    seed = int(config["seed"])
    batch_size = int(config["batch_size"])

    metric = evaluate.load("glue", "mrpc")

    # If the batch size is too big we use gradient accumulation
    gradient_accumulation_steps = 1

    set_seed(seed)
    train_dataloader, eval_dataloader = get_dataloaders(batch_size)
    model = AutoModelForSequenceClassification.from_pretrained("bert-base-cased", return_dict=True)
    model = model.to(device)

    # Instantiate optimizer
    optimizer = AdamW(params=model.parameters(), lr=lr)

    # Instantiate scheduler
    lr_scheduler = get_linear_schedule_with_warmup(
        optimizer=optimizer,
        num_warmup_steps=100,
        num_training_steps=(len(train_dataloader) * num_epochs) // gradient_accumulation_steps,
    )

    for epoch in range(num_epochs):
        model.train()
        for step, batch in enumerate(train_dataloader):
            batch.to(device)
            outputs = model(**batch)
            loss = outputs.loss
            loss = loss / gradient_accumulation_steps
            loss.backward()
            if step % gradient_accumulation_steps == 0:
                optimizer.step()
                lr_scheduler.step()
                optimizer.zero_grad()

        model.eval()
        samples_seen = 0
        for step, batch in enumerate(eval_dataloader):
            batch.to(device)
            with torch.no_grad():
                outputs = model(**batch)
            predictions = outputs.logits.argmax(dim=-1)
            metric.add_batch(
                predictions=predictions,
                references=batch["labels"],
            )

        eval_metric = metric.compute()
        # Use accelerator.print to print only on the main process.
        print(f"epoch {epoch}:", eval_metric)


def main(device):
    config = {"lr": 2e-5, "num_epochs": 3, "seed": 42, "batch_size": 16}
    training_function(config, device)

from time import time

for device in ["cpu", "mps"]:
    device = torch.device(device)
    print(device)
    start_time = time()
    main(device)
    end_time = time()
    print(f"time taken in seconds for training using device {device}: {end_time-start_time}")
```
2. The output logs:
```bash
CPU
...

epoch 0: {'accuracy': 0.7696078431372549, 'f1': 0.8540372670807453}
epoch 1: {'accuracy': 0.8357843137254902, 'f1': 0.8858603066439524}
epoch 2: {'accuracy': 0.8480392156862745, 'f1': 0.8934707903780068}
time taken in seconds for training using device cpu: 1001.8208646774292

mps

...

epoch 0: {'accuracy': 0.6838235294117647, 'f1': 0.8122270742358079}
epoch 1: {'accuracy': 0.7058823529411765, 'f1': 0.8214285714285715}
epoch 2: {'accuracy': 0.6985294117647058, 'f1': 0.8183161004431315}
time taken in seconds for training using device mps: 771.7576680183411
```
# Expected behaviour:
Same performance metrics when using `mps` device in comparison to `cpu` training.

### Versions

Collecting environment information...
PyTorch version: 1.12.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.5 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.10.5 (v3.10.5:f377153967, Jun  6 2022, 12:36:10) [Clang 13.0.0 (clang-1300.0.29.30)] (64-bit runtime)
Python platform: macOS-12.5-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.1
[pip3] torch==1.12.0
[pip3] torchaudio==0.12.0
[pip3] torchvision==0.13.0
[conda] Could not collect

cc @VitalyFedyunin @ngimel @kulinseth @albanD

**Comment by philipturner** [comment]
This is related to https://github.com/pytorch/pytorch/issues/77799. I suspect it's because of overhead of using `MPSGraph` for everything. On the Apple M1 Max, there is:

- 10 µs overhead to create a new `MTLCommandBuffer` for each op
- 15 µs overhead to encode the `MPSGraph` for each op, if it's already compiled into an `MPSGraphExecutable`. This doesn't change even if you put multiple ops into the `MPSGraph`; overhead just becomes (15 µs) x (#ops)
- 100 µs (I need to recreate the benchmark that proved this) overhead to encode an `MPSGraph` if it isn't already compiled into an `MPSGraphExecutable`

That means 25 - 110 µs of overhead for a single operation, like the `negate` operator. Multiply that by 100s of thousands, and you get many, many milliseconds of extra execution time. This overhead can be avoided and brought down to 0.5 µs (encode time) / 3.0 µs (GPU sequential throughput), but you need to use custom Metal shaders for trivial ops instead of a new `MPSGraph` instance for each op.

On the CPU, everything is instantaneous. Overhead of calling into something might be 25 - 110 _nanoseconds_, not 25 - 110 _microseconds_. RNNs like BERT have smaller tensors and more sequential in nature, so they often run better on CPUs - which have higher clock speed and smaller vector width. Perhaps this would not be true if a GPU had overhead of < 1 microseconds.

More information is in the prototype repository [here](https://github.com/philipturner/metal-experiment-1), which applies several unique optimizations to drive down overhead. It will eventually migrate to [s4tf/metal](https://github.com/s4tf/metal).

**Comment by pacman100** [comment]
Speedup is important but I am more concerned about performance metrics `accuracy` and `F2` scores. Performance drop of -17% in `Accuracy` and -9% in `F1` score compared to CPU only training when there are no code changes expect for using `mps` device.

**Comment by pacman100** [comment]
For more context, inference of a trained/finetuned transformer model seems to give expected performance metrics. Hence, training is going haywire while using `mps` device.

**Comment by philipturner** [comment]
Is it possible to unravel the BERT model's layers into a sequence of primitive operations? Then, perhaps I could replicate the results from scratch using the `MPSGraph` framework. I'm using an M1-family Mac as well, so numerical precision should be exactly the same.

**Comment by DenisVieriu97** [comment]
@pacman100 could you please try latest pytorch nightly (such as 1.13.0.dev20220805) and let me know if you are still seeing the issue? (`pip3 install --force-reinstall --pre torch --extra-index-url https://download.pytorch.org/whl/nightly/cpu` will bring the latest nightly version)

**Comment by pacman100** [comment]
Thank you @philipturner for the detailed insights into what might be happening behind the scenes. 
Thank you @DenisVieriu97 for the suggestion to test out `mps` support using PyTorch nightly.

Observations:
1. With PyTorch nightly, the performance is similar (same for the first 2 decimal points) (-0.3% F1 drop and -0.6% Accuracy drop) as seen below.  **Therefore, model correctness/performance metrics seem to be resolved.**
2.  We can also observe **~60%** speedup compared to the ~30% speedup from the torch 1.12.0 version. However, this is nowhere near the 10X eval speedup for Bert mentioned in the blog [Introducing Accelerated PyTorch Training on Mac | PyTorch](https://pytorch.org/blog/introducing-accelerated-pytorch-training-on-mac/). To provide more context, I observed a speedup of ~7.5X over CPU for the task of image classification using ResNet which is in line with the blog post. **Therefore, now the pressing question is of performance/speedup for transformer-based models.**

```bash
cpu
epoch 0: {'accuracy': 0.7696078431372549, 'f1': 0.8540372670807453}
epoch 1: {'accuracy': 0.8357843137254902, 'f1': 0.8858603066439524}
epoch 2: {'accuracy': 0.8480392156862745, 'f1': 0.8934707903780068}
time taken in seconds for training using device cpu: 982.4634261131287

...


mps
epoch 0: {'accuracy': 0.7475490196078431, 'f1': 0.8422664624808577}
epoch 1: {'accuracy': 0.8431372549019608, 'f1': 0.8810408921933086}
epoch 2: {'accuracy': 0.8431372549019608, 'f1': 0.8900343642611684}
time taken in seconds for training using device mps: 608.2880346775055
```

**Comment by philipturner** [comment]
> **Therefore, now the pressing question is of performance/speedup for transformer-based models.**

I've been waiting for this chance for a very long time. I wanted a real-world use case (specifically an RNN) to showcase measurable speedups from reducing driver-side overhead. If you could humor me, I'd like to research the performance regression by benchmarking PyTorch against the rebooted S4TF. Please let me know whether you're interested in helping with this investigation.

<details>
<summary>Stage 1</summary>

My custom backend isn't yet in the state where it can test this BERT model. There's still quite a bit of work and debugging to complete. However, lowering it down to `MPSGraph` is a good start. This is something called narrowing a bug - eliding as much context as possible, to remove confounding variables and examine only the root cause.

Despite making backends for ML frameworks, I'm not an expert on using ML frameworks. I have never used S4TF, PyTorch, or TensorFlow to train an actual model, except when running a pre-created tutorial notebook on [Swift-Colab](https://github.com/philipturner/swift-colab). I care more about making the software accessible to other people, who want to do training. So I may need help understanding or extracting to model's size parameters, sequence of raw ops, etc.
</details>

<details>
<summary>Stage 2</summary>

Investigate ways to run the model faster than one `MPSGraph` per op. Try fusing hotpaths or creating custom Metal shaders, just for demonstration. These emulate the final product of my prototype Metal backend, and may signal whether my 10-100x speedup hypothesis is reasonable. Then, see whether such optimizations are possible in PyTorch. Odds are, they aren't. There's something unique about Swift, how it handles ARC and does automatic differentiation in the language compiler.

There's also the high maintenance cost of my highly tuned command stream + memory management. It requires writing _lots_ of low-level code in a safe, productive programming language. It's counterproductive to translate it into C++, or worse, Objective-C (Swift's predecessor). Still, this investigation could reveal limitations of PyTorch's MPS backend, including _reproducible_ evidence of why MPS is sometimes slower than the CPU.
</details>

**Comment by leovinus2001** [comment]
> > **Therefore, now the pressing question is of performance/speedup for transformer-based models.**
> 
> I've been waiting for this chance for a very long time. 

Apologies if you all know and have digested this a while back but related to speedup of Transformers and RNNs on Apple devices, specifically ANE, there are lessons and code to be learned in two links below. Obviously that is yet-another-backend compared to CPU and MPS+GPU. The data layout discussion might be relevant for MPS speedup while the 4D format for convolutions seems very ANE specific. Just some additional background info for Transformer speedup on devices.
https://github.com/apple/ml-ane-transformers
https://machinelearning.apple.com/research/neural-engine-transformers

**Comment by alin256** [comment]
> don't handle infinities and denormals

I am getting **infinite loss** in my project when enabling **MPS**, which never occurs in either CPU or CUDA.
Can this be related to what @philipturner describes?
It is a bit hard to isolate the issue, unfortunately.

**Comment by philipturner** [comment]
The project I'm making always uses the `precise::` namespace. This isn't currently possible in PyTorch, which uses `MPSGraph` for activations. The biggest bottleneck is memory bandwidth, not ALU time - clock cycles are basically free. There's no reason to avoid `precise::`. 

@alin256 could you show some of the source code that's causing infinite loss? If so, I might be able to trace back the culprit and show how using `precise::` would change results.

**Comment by alin256** [comment]
I looked deeper, and something really dodgy was going on:

My input tensors' means on GPU are nans

```
            data_input1, data_input2, drill_traj, data_output = data

            print("inputs cpu")
            print(torch.mean(data_input1.float()))
            print(torch.mean(data_input2.float()))
            print(torch.mean(drill_traj.float()))

            print("inputs gpu")
            data_input1 = data_input1.float().to(device=device, non_blocking=True)
            data_input2 = data_input2.float().to(device=device, non_blocking=True)
            data_input3_drill = drill_traj.float().to(device=device, non_blocking=True)
            data_output = data_output.float().to(device=device, non_blocking=True)
```

```
inputs cpu
tensor(0.1111)
tensor(0.1129)
tensor(-1.0027)
inputs gpu
/opt/anaconda3/envs/multi-modal-m1/lib/python3.8/site-packages/torch/_tensor_str.py:103: UserWarning: The operator 'aten::bitwise_and.Tensor_out' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at  /Users/runner/work/_temp/anaconda/conda-bld/pytorch_1659484780698/work/aten/src/ATen/mps/MPSFallback.mm:11.)
  nonzero_finite_vals = torch.masked_select(tensor_view, torch.isfinite(tensor_view) & tensor_view.ne(0))
tensor(0.1111, device='mps:0')
tensor(nan, device='mps:0')
tensor(nan, device='mps:0')
```

**Comment by philipturner** [comment]
Could you extract the raw data of one of those tensors before computing the mean? We can start by putting it into a calculator or simple script and validating that the CPU has it right. Understanding what values trigger the `NAN` will help investigate the GPU problems.

I'm hoping for the three raw data's that cause:
```
tensor(0.1111)
tensor(0.1129) vs. tensor(nan)
tensor(-1.0027) vs. tensor(nan)
```

Use a `<details> </details>` if the data is quite large, otherwise you will flood the thread.

**Comment by kulinseth** [comment]
> 1. With PyTorch nightly, the performance is similar (same for the first 2 decimal points) (-0.3% F1 drop and -0.6% Accuracy drop) as seen below.  **Therefore, model correctness/performance metrics seem to be resolved.**
> 2. We can also observe **~60%** speedup compared to the ~30% speedup from the torch 1.12.0 version. However, this is nowhere near the 10X eval speedup for Bert mentioned in the blog [Introducing Accelerated PyTorch Training on Mac | PyTorch](https://pytorch.org/blog/introducing-accelerated-pytorch-training-on-mac/). To provide more context, I observed a speedup of ~7.5X over CPU for the task of image classification using ResNet which is in line with the blog post. **Therefore, now the pressing question is of performance/speedup for transformer-based models.**

Thanks @pacman100 for trying out the torch-nightly. I will consider the Correctness issue fixed with latest torch nightly. For the performance I would like to provide more context. The numbers in the Blog post were collected on M1 Ultra machine which has 4x the number of GPU cores and 128GB of unified memory. Also the benchmark used was the `torchbench` benchmark `hf_Bert` model at the specified Batch sizes. 
<img width="863" alt="Screen Shot 2022-08-08 at 10 19 53 AM" src="https://user-images.githubusercontent.com/453004/183476158-1be558a1-f961-4d89-bfb0-4b647fe6bea1.png">

We are continuing to make perf improvements so stay tuned on that.

**Comment by kulinseth** [comment]
> > don't handle infinities and denormals
> 
> I am getting **infinite loss** in my project when enabling **MPS**, which never occurs in either CPU or CUDA. Can this be related to what @philipturner describes? It is a bit hard to isolate the issue, unfortunately.

@alin256 , can you please file a separate issue on this and we will take a look. Thanks!

**Comment by philipturner** [comment]
> @alin256 , can you please file a separate issue on this and we will take a look. Thanks!

Please CC me as well.

@kulinseth we were still discussing performance issues on this thread, and a potential investigation. I don't think it's correct to mark it as closed on GitHub.

**Comment by kulinseth** [comment]
> @kulinseth we were still discussing performance issues on this thread, and a potential investigation. I don't think it's correct to mark it as closed on GitHub.

The bug was tracking the `accuracy` and `f1` performance metrics in the BERT network, which was addressed. The other point on Performance was raised about not matching the Blogpost which I provided more clarification. If this is not satisfactory @pacman100 , feel free to reopen the issue.

About discussing the performance issues on this thread, I would appreciate we create a specific issue and track it there. This issue is getting conflated with parallel threads.

**Comment by AlexanderTekle** [comment]
(Bumping up this old bug, let me know if there's a better place)

I'm also noticing that mps produces consistently lower results than what model is trained on cpu. However, the results are much smaller than OP's original scenar, are are usually maybe -.02 lower. There's an argument for this to be within the room for error, however I'm noticing this across runs extremely consistently (failed to see one where mps converges on a higher f1 score)

Mps 
epoch 0: {'accuracy': 0.7769607843137255, 'f1': 0.8330275229357799}
epoch 1: {'accuracy': 0.8406862745098039, 'f1': 0.8873483535528597}
epoch 2: {'accuracy': 0.8382352941176471, 'f1': 0.8862068965517241}
time taken in seconds for training using device mps: 576.6726970672607

cpu
epoch 0: {'accuracy': 0.7573529411764706, 'f1': 0.8070175438596492}
epoch 1: {'accuracy': 0.8504901960784313, 'f1': 0.8924162257495591}
epoch 2: {'accuracy': 0.8676470588235294, 'f1': 0.9049295774647886}
time taken in seconds for training using device cpu: 1144.207025051117
